v2/azure-shared-resource.go

```diff
@@ -95,94 +115,71 @@ func (r *azureSharedResource) WithFactor(val uint32) AzureSharedResource {
-	r.reservedCapacity = val
+	r.phaseMutex.Lock()
+	defer r.phaseMutex.Unlock()
+	if r.phase != batcherPhaseUninitialized {
+		panic(InitializationOnlyError)
+	}
+	atomic.StoreUint32(&r.reservedCapacity, val)
 	return r
 }
 
 // The rate limiter will attempt to obtain an exclusive lease on a partition (when needed) every so often. The interval is random to
 // reduce the number of collisions and to provide an equal opportunity for processes to compete for partitions. This setting determines
 // the maximum amount of time between intervals. It defaults to `500` and is measured in milliseconds.
 func (r *azureSharedResource) WithMaxInterval(val uint32) AzureSharedResource {
+	r.phaseMutex.Lock()
+	defer r.phaseMutex.Unlock()
+	if r.phase != batcherPhaseUninitialized {
+		panic(InitializationOnlyError)
+	}
 	r.maxInterval = val
 	return r
 }
 
-// DEPRECATED
-func (r *azureSharedResource) Provision(ctx context.Context) (err error) {
-
-	// only allow one phase at a time
+// You may provide a lease time in seconds. This defaults to 15 seconds. This is passed to the lease call made in Azure provided it is between
+// 15 and 60 seconds. If you provide a lower value, 15 seconds will be used. If you provide a greater value, 60 seconds will be used. If you are
+// using WithMocks, this constraint on the time will the ignored.
+func (r *azureSharedResource) WithLeaseTime(val uint32) AzureSharedResource {
 	r.phaseMutex.Lock()
 	defer r.phaseMutex.Unlock()
-	if r.phase != rateLimiterPhaseUninitialized {
-		err = ImproperOrderError
-		return
+	if r.phase != batcherPhaseUninitialized {
+		panic(InitializationOnlyError)
 	}
-
-	// check requirements
-	if r.leaseManager == nil {
-		err = UndefinedLeaseManagerError
-		return
-	}
-	if r.factor == 0 {
-		r.factor = 1 // assume 1:1
-	}
-	if r.sharedCapacity < 1 {
-		err = UndefinedSharedCapacityError
-		return
-	}
-	if r.maxInterval < 1 {
-		r.maxInterval = 500 // default to 500 ms
-	}
-
-	// get a write lock on partitions
-	r.partlock.Lock()
-	defer r.partlock.Unlock()
-
-	// provision the container
-	err = r.leaseManager.provision(ctx)
-	if err != nil {
-		return
-	}
-
-	// make 1 partition per factor
-	count := int(math.Ceil(float64(r.sharedCapacity) / float64(r.factor)))
-	if count > 500 {
-		err = PartitionsOutOfRangeError{
-			MaxCapacity:    r.MaxCapacity(),
-			Factor:         r.factor,
-			PartitionCount: count,
-		}
-		return
-	}
-	r.partitions = make([]*string, count)
-
-	// provision partitions
-	err = r.leaseManager.createPartitions(ctx, count)
-
-	// mark provision as completed
-	r.phase = rateLimiterPhaseProvisioned
-
-	return
+	r.leaseTime = val
+	return r
 }
 
 // This returns the maximum capacity that could ever be obtained by the rate limiter. It is `SharedCapacity + ReservedCapacity`.
 func (r *azureSharedResource) MaxCapacity() uint32 {
-	return r.sharedCapacity + r.reservedCapacity
+	return atomic.LoadUint32(&r.sharedCapacity) + atomic.LoadUint32(&r.reservedCapacity)
```

Comment:
Since `provisionBlobs` potentially caps capacity if (sharedCapacity / factor) > 500, should this also reflect that possible capping? I would hope that users are handling that error and it wouldn't get to this, but it seems possible that it could (in which case the batcher might also allow enqueue even though it shouldn't).

Your comments on `WithFactor` speaks to this and an error is emitted when that condition is hit in the function, but I suppose it's a question of if we want to reflect the *actual* or specified/expected value here.

Reply:
This is a great callout. I did not even consider that, but it seems like it should be the maximum that could be obtained, which would also help if someone didn't look for the error event.

I have changed it to respect the 500 partition cap.